### ISSUE TITLE
ErrNotFound instead of Duplicate Error on upsert breaks compatibility

### ISSUE THREAD

**Issue by OSadovy** [go-mgo/mgo#313]
Hi,

Part of our app uses mongo db for distributed lock, which broke with recent build of mgo/v2. I nailed down the problem to merge of pull request #291 .

Example:
```
func mongoLock() (bool, string, error) {
	dt := time.Now().UTC()
	var (
		lock         bson.M
		prevInstance string
	)
	_, err := rm.db.C("systemlocks").Find(
		bson.M{"_id": "all", "dt": bson.M{"$lt": dt.Add(-LockInterval)}},
	).Apply(mgo.Change{
		Update: bson.M{"$set": bson.M{"dt": dt, "instance": InstanceName}},
		Upsert: true,
	}, &lock)
	if err != nil {
		if mgo.IsDup(err) {
			return false, "", nil
		}
		return false, "", err
	}

	if _, ok := lock["instance"]; ok {
		prevInstance = lock["instance"].(string)
	}
	return true, prevInstance, nil
}

func mongoUnlock() error {
	dt := time.Now().UTC()
	err := rm.db.C("systemlocks").Update(
		bson.M{"_id": "all", "instance": InstanceName, "dt": bson.M{"$lte": dt}},
		bson.M{"$set": bson.M{"dt": time.Time{}}},
	)
	if err != nil {
		if err == mgo.ErrNotFound {
			log.Error("Error releasing lock: not acquired!")
			return nil
		}
		return err
	}
	return nil
}
```

Earlier mgo returned duplicate error when another instance held the lock, but now, I assume it retries and returns ErrNotFound. Reading mongo documentation which was referenced in the above pull request, it says "If the operation fails because of a duplicate index key error, applications may retry the operation which will succeed as an update operation."

*May* is not *should*, so maybe you reconsider the decision to retry *every* upsert operation silently?

**Comment by niemeyer** [comment]
That's certainly interesting. Can you provide more details about why you are trying to catch an ErrNotFound on an upsert, given that the typical semantics is "insert if it does not exist", rather than returning "not found"?

That's the key of the problem. It is indeed incompatible with the previous behavior, but my reading is that this was actually a bug fix for pretty much every application relying on mgo today. I've never seen myself people expecting that error on their upserts. Typically it just breaks down.

**Comment by OSadovy** [comment]
@niemeyer,

Actually, as per my code example, I am expecting dup error, not ErrNotFound.
```
if mgo.IsDup(err) {
return false, "", nil // false means lock is not acquired because it is held by another instance
}
```

ErrNotFound is that my app is getting after build with latest changes of mgo driver. That breaks compatibility.

**Comment by niemeyer** [comment]
The question and the point are the same. Please just replace ErrNotFound by "duplicated error" in the above comment. What you are relying on is undocumented behavior, which will break most applications as they expect an Update+Insert operation to actually _update_ if it exists.

Can you please detail why is the previous behavior appropriate for your application, and how is it now breaking it?

Thanks again for your help on this issue.

**Comment by cezarsa** [comment]
I see two problems with the recent change in behavior. First is that the error returned by calling `Upsert()` is not consistent with the error returned by `Find().Apply({Upsert: true})`. Full example of this is in https://gist.github.com/cezarsa/1a244965ae085c919336c58815172e49. But basically assuming a record `{_id: 1, x: 1}` exists, doing `c.Upsert(bson.M{"_id": 1, "x": 2}, bson.M{"$set": bson.M{"x": 3}})` will cause a `mgo.IsDup()` error to be returned, the same as before. However, calling ```c.Find(bson.M{"_id": 1, "x": 2}).Apply(mgo.Change{
		Update: bson.M{"$set": bson.M{"x": 3}},
		Upsert: true,
	}, nil)```

**Comment by cezarsa** [comment]
Sorry I erased my previous comment because I hit send by accident before finishing.

I can se two different problems with the change in behavior. The first is that the error returned by `Upsert()` will not be consistent with the error returned by `Find().Apply({Upsert: true})`. A full example of this is in https://gist.github.com/cezarsa/1a244965ae085c919336c58815172e49.

Basically assuming a record with `{_id: 1, x: 1}` exists. This:
```
c.Upsert(bson.M{"_id": 1, "x": 2}, bson.M{"$set": bson.M{"x": 3}})
```
will return a `mgo.IsDup()` error just as before, where this:

```
c.Find(bson.M{"_id": 1, "x": 2}).Apply(mgo.Change{
	Update: bson.M{"$set": bson.M{"x": 3}},
	Upsert: true,
}, nil)
```
will return a not found error.


The other problem I can see is that the behavior `mgo` and the mongodb console is also not consistent.

Executing the following `findAndModify` command on console will raise a duplicate key error:
```
$ mongo
MongoDB shell version: 3.0.12
connecting to: test
> db.a.insert({_id: 1, x: 1})
WriteResult({ "nInserted" : 1 })
> db.a.findAndModify({query: {_id: 1, x: 2}, update: {$set: {x: 3}}, upsert: true})
2016-08-11T18:11:10.120+0000 E QUERY    Error: findAndModifyFailed failed: {
	"errmsg" : "exception: E11000 duplicate key error index: test.a.$_id_ dup key: { : 1.0 }",
	"code" : 11000,
	"ok" : 0
}
    at Error (<anonymous>)
    at DBCollection.findAndModify (src/mongo/shell/collection.js:636:15)
    at (shell):1:6 at src/mongo/shell/collection.js:636
```

The same dup key error will be returned for a `update({}, {}, {upsert: true)` command.

**Comment by niemeyer** [comment]
@cezarsa That looks like an actual bug in the new logic inside Apply. Will fix it, thanks for raising it.

With that aside, do you depend on the previous behavior for some reason, or is the automatic retrial helpful?

**Comment by cezarsa** [comment]
Thanks for the fast fix. I have some code similar to that of @OSadovy for distributed locks, however most of it was using `Upsert()` so I wasn't much affected. For the automatic retry on dup it is indeed useful, much better than having to retry it on my side, especially with mongo 3.2 where these kind of errors seem to happen a lot more. As long as a dup error is returned after N retries (as seems to be the case in mgo) I won't have to change any of my code.

**Comment by OSadovy** [comment]
@niemeyer,

The idea of my lock implementation is based on the fact, that either find part of the operation succeeds (lock is released / has been timed out) and modify part will acquire it; or
find fails because lock is held by another app instance ('dt' field is greather than current time), and modify tries to upsert a document and fail because document with same _id already exists, and my app will act accordingly - eg: explicitly retries in some interval, or moves on.

I think that mgo implicit automatic retrying will not help because the knowledge how/when to do it is very app specific.

Thanks @cezarsa  for your example - that is exactly my problem!

**Comment by niemeyer** [comment]
#316 fixes the problem reported by @cezarsa, and I understood what you said above, the retrying won't actually be a problem for you once that fix lands. Your lock will still work fine.

Retrying after an upsert is actually not app specific, which is what motivated this fix. Pretty much every single "normal" application (not leveraging edge behavior for alternative meanings) I've seen using Upsert and FindAndModify with Upsert set did not expect an IsDup out of it. These simply want to retry, as @cezarsa pointed out above.

**Comment by niemeyer** [comment]
Fixed with #316.
